rootston/seat.c

```diff
@@ -613,6 +613,9 @@ static void handle_keyboard_destroy(struct wl_listener *listener, void *data) {
 	wl_list_remove(&keyboard->device_destroy.link);
 	wl_list_remove(&keyboard->keyboard_key.link);
 	wl_list_remove(&keyboard->keyboard_modifiers.link);
+	if (seat->seat->keyboard_state.keyboard == keyboard->device->keyboard) {
+		wlr_seat_set_keyboard(seat->seat, NULL);
```

Comment:
>Because I haven't seen any code possibly sending NULL to seat_set_keyboard, it seems that an active keyboard being destroyed would have never cleared the problematic listeners, even with other keyboard types. That's why I added the fix to the general handler in seat.

Hmm, we already set it to NULL in `handle_keyboard_destroy` in `seat/wlr_seat_keyboard.c`. I don't understand why this is needed.

Reply:
That's because `wlr_sear_set_keyboard` also removes listeners pointing to signals on `wlr_input_device` in `seat/wlr_seat_keyboard.c:127`. Those signals become invalid after destruction, so the disconnection must happen beforehand.

If I'm reading this right, the listeners are attached to the seat, so they will keep references even after referenced keyboards are gone.

Reply:
Maybe calling `wlr_seat_set_keyboard(seat, NULL)` in `handle_keyboard_destroy` is the right fix?

Reply:
That's what's happening here - note that I moved the call.

Reply:
But this is in rootston - not in wlroots.